python/felis/db/database_context.py

```diff
@@ -0,0 +1,859 @@
+    """Interface for managing database operations across different
+    SQL dialects.
+    """
+
+    @property
+    @abstractmethod
+    def metadata(self) -> MetaData:
+        """The SQLAlchemy metadata representing the database for the context
+        (`~sqlalchemy.sql.schema.MetaData`).
+        """
+        ...
+
+    @property
+    @abstractmethod
+    def engine(self) -> Engine:
+        """The SQAlchemy engine for the context (`~sqlalchemy.engine.Engine`).
+
+        Raises
+        ------
+        DatabaseContextError
+            If an engine is not available for this context.
+        """
+        ...
+
+    @property
+    @abstractmethod
+    def dialect(self) -> Dialect:
+        """The SQLAlchemy dialect for the context
+        (`~sqlalchemy.engine.Dialect`).
+        """
+        ...
+
+    @property
+    @abstractmethod
+    def dialect_name(self) -> str:
+        """Get the dialect name for this database context."""
+        ...
+
+    @abstractmethod
+    def initialize(self) -> None:
+        """Create the target schema in the database if it does not exist
+        already.
+
+        Raises
+        ------
+        DatabaseContextError
+            If there is an error instantiating the schema.
+        """
+        ...
+
+    @abstractmethod
+    def drop(self) -> None:
+        """Drop the schema in the database if it exists.
+
+        Raises
+        ------
+        DatabaseContextError
+            If there is an error dropping the schema.
+        """
+        ...
+
+    @abstractmethod
+    def create_all(self) -> None:
+        """Create all database objects in the schema using the metadata
+        object.
+
+        Raises
+        ------
+        DatabaseContextError
+            If there is an error creating the schema objects in the database.
+        """
+        ...
+
+    @abstractmethod
+    def create_indexes(self) -> None:
+        """Create all indexes in the schema using the metadata object.
+
+        Raises
+        ------
+        DatabaseContextError
+            If there is an error creating the indexes in the database.
+        """
+        ...
+
+    @abstractmethod
+    def drop_indexes(self) -> None:
+        """Drop all indexes in the schema using the metadata object.
+
+        Raises
+        ------
+        DatabaseContextError
+            If there is an error dropping the indexes in the database.
+        """
+        ...
+
+    @abstractmethod
+    def execute(self, statement: SQLStatement) -> Result:
+        """Execute a SQL statement and return the result.
+
+        Parameters
+        ----------
+        statement
+            The SQL statement to execute.
+
+        Returns
+        -------
+        `~sqlalchemy.engine.Result`
+            The result of the statement execution.
+
+        Raises
+        ------
+        DatabaseContextError
+            If there is an error executing the SQL statement.
+        """
+        ...
+
+
+class _BaseContext(DatabaseContext):
```

Comment:
This should probably be abstract (ABC) as well?
And if so you'd have to add the abstract methods (initialize & drop)

Reply:
An inherited abstract class is still considered abstract unless it implements all abstract methods, so I think it will be sufficient just to make `DatabaseContext` an `ABC`.